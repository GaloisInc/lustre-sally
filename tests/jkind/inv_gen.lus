-- This file contains a simple property which requires the invariant
-- described by 'lemma' (either explicitly or through invariant
-- generation)

node main() returns (x: bool);
var
   count: int; 
   ok: bool;
   lemma: bool;
let
   count = 1 -> 
      (if (pre(count) < 5)
         then pre(count) + 1
         else 5);

   x = false -> ((not (count >= 5)) and (pre x));

   ok = true -> ((not (pre x)) or x);
   --%PROPERTY ok;

   lemma = not x;
   -- %PROPERTY lemma;
tel;
