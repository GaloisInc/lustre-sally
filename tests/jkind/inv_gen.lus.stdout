Sally Model
===========

(define-state-type
   S
   (
     (count Int)
     (lemma Bool)
     (ok Bool)
     (x Bool)
     (|count:init| Bool)
     (|gal-initializing| Bool)
     (|ok:init| Bool)
     (|pre_count:1| Int)
     (|pre_count| Int)
     (|pre_x:1| Bool)
     (|pre_x| Bool)
     (|x:init| Bool))
   ())
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |count:init| true)
      (= |x:init| true)
      (= |ok:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|pre_count:1| state.count)
      (= next.|pre_count| state.count)
      (ite
         state.|count:init|
         (= next.count 1)
         (=
            next.count
            (ite
               (< next.|pre_count:1| 5)
               (+ next.|pre_count| 1)
               5)))
      (= next.|count:init| false)
      (= next.|pre_x:1| state.x)
      (ite
         state.|x:init|
         (= next.x false)
         (=
            next.x
            (and
               (not
                  (<= 5 next.count))
               next.|pre_x:1|)))
      (= next.|x:init| false)
      (= next.|pre_x| state.x)
      (ite
         state.|ok:init|
         (= next.ok true)
         (=
            next.ok
            (or
               (not next.|pre_x|)
               next.x)))
      (= next.|ok:init| false)
      (=
         next.lemma
         (not next.x))))

(query
   TS
   (or |gal-initializing| ok))

[Lustre] Validating properties:
[Lustre] Property ok... [Unknown] Valid up to depth 10
[Lustre] Model status: [Unknown] 
